posts/2020/07/the-ibis-backends.md

```diff
@@ -0,0 +1,351 @@
+We'll recognize the __execute__ method when we return `pandas.DataFrame`s from `ibis` expression.
+The __compile__ method does not trigger any computation, rather it constructs an intermediate form
+that is interpretted by a backend.
+
+```python
+    >>> assert isinstance(expression.compile(), ibis.expr.types.TableExpr)
+```
+
+In the case of direction execution backends, the `expression` compiles to an the original `ibis` 
+expression.  And the computation is carried out based on a set of recipes defined in `ibis`.
+
+In general, we would typically do this work directly in `pandas`, however this work is
+practical in mocking tests for expressions independent of backends.
+
+> Learn more about the [HDF5 direct execution backend in the `ibis` tests][test-hdf5].
+
+## `ibis` expression generating backends.
+
+```python
+    db = ibis.sqlite.connect('lahmansbaseballdb.sqlite')
+    expression = db.table('halloffame').head()
+```
+
+Expression generating backends operate on [SQL] databases that interoperate with `sqlalchemy`.
+
+```python
+    >>> assert isinstance(expression.compile(), sqlalchemy.sql.Select)
+```
+
+In the case of expression generating backends, the intermediate representation is a `sqlalchemy` object.
+`sqlalchemy` is _The Database Toolkit for Python_, and `ibis` leverages it compatability
+with traditional [SQL] databases.
+    
+
+
+## `ibis` string generating backends.
+
+```bash
+pip install --upgrade ibis-framework[omniscidb]
+conda install -c conda-forge ibis-framework # install all the backends!
+```
+
+String generating backends allow `ibis` to interface with big data systems that manage 
+their own computation. For example, we may connect to an example `omnisci` database.
+    
+
+```python
+    import ibis.omniscidb
+    omnisci = ibis.omniscidb.connect(host='metis.omnisci.com', user='demouser', password='HyperInteractive', port=443, database='omnisci', protocol='https')
+```
+    
+`omnisci` is described as a string generating backend because the intermediate representation of the
+query is a flavor of SQL.
+
+
+```python
+    expression = omnisci.table('upstream_reservoir').head()
+```
+
+    
+A string generating expression compiles to `ibis.omniscidb` flavored [SQL], while `ibis.bigquery` may have a different string representation.
+
+
+```python
+    >>> expression.compile()
+    'SELECT *\nFROM upstream_reservoir\nLIMIT 5'
+```
+    
+
+> Major credit goes to [@xmnlab] in his heroic PR to introduce `ibis.omniscidb` into `ibis`. You can watch
+the drama play out in this [Github Issue][omnisci-pr]. If you'd like to learn more about [OmniSci] and
+`ibis.omniscidb` checkout the following links.
+> * [OmniSci][omnisci]
+> * [Quansight Labs- Ibis: Python data analysis productivity framework][labs-post]
+
+## Conclusion
+
+We'd like to thank the maintainers of the `ibis` for
+their and effort in supporting the `ibis` community.
+
+
+[sql-server]: #
```

Comment:
ah I now see where the links are... so many of my above links comments may be invalid. But it seems that some of these are still empty.